File: R/IFL_Algorithm.R
#
#  IFL: Iterative Fused LASSO Algorithm
#
#  Last update: 19/07/25
#

library(glmnet)     # LASSO and Elastic-Net 

# Supporting Function
source("supportingfunctions.R")

#
# IFL
#

# IFL(y, x, handle_outlier = true, intercept = true)
# 
# Intercept = true => There is an intercept in x, then once the data is centralized, 
#   it can be removed from the first column of X (zeros) 
# 
# Returns beta_np_1 = v_beta_Par , beta_n_p = m_beta_Par,           # Solutions without Outliers
#         betao_np_1 = v_beta_Paro , betao_n_p = m_beta_Paro,       # Solutions with Outliers
#         beta0 = beta0_hat,                                        # Estimated Intercept
#         y_hat =  y_original_hat ,                                 # Insample Estimators without Outliers
#         y_hato = y_original_hat_woutl,                            # Insample Estimators with Outliers
#         mean_x  = mean_x , mean_y  = mean_y,                      # column-wise mean of X and mean of y
#         bic_value = bic_value,                                    # BIC value model without outliers
#         bic_value_outliers = bic_value_outliers,                  # BIC value model with outliers
#         chosen_var = chosen_variables,                            # List of selected columns in X
#         chosen_var_bk = chosen_variables_break,                   # List of selected columns with estim breaks in X
#         chosen_var_outliers = chosen_variables_outliers,          # List of selected columns in X (model w/outliers)
#         chosen_var_outliers_bk = chosen_variables_outliers_break, # List of selected columns with estim breaks in X (model w/outliers)
#         RMSE_inS = RMSE_inS, RMSE_inSo = RMSE_inSo,               # In Sample RMSE (models with and without outliers)
#         ind_out = ind_out,                                        # list of outlier in y  
#         IFL_t0 = IFL_t0                                           # time to converge
# 
# Last review: 19 Jul 25
#

IFL <- function(y, x, handle_outlier = TRUE, intercept = TRUE) {

  
  # TESTE
  #handle_outlier <- TRUE
  #intercept <- TRUE
  # TESTE
  
  # Store originals
  x_original <- x
  y_original <- y
  
  # Always center the data
  mean_x <- colMeans(x_original) 
  mean_y <- mean(y_original)
  beta0 = mean_y
  y <- y_original - mean_y
  x <- scale(x_original, center = TRUE, scale = FALSE)
  
  # No need to process intercept column that equals 0!
  if (intercept) {
    x <- x[, -1]
  }
  
  ind_out <- numeric()
  
  ########################################################
  # Problem Dimensions
  n <- length(y) 
  p <- dim(x)[2]
  ########################################################
  
  # Create Matrix XX 
  XX <- NULL
  for (i in 1:p) {
    #i <- 1
    XX <- cbind(XX, diag(x[,i]))
  }
  
  # Setup Iteractive Fused LASSO 
  nbetas_In  <- c(rep(n,p))
  nbetas_In_Tot <- sum(nbetas_In)
  nbetas_Out <- c(rep(0,p))
  nbetas_Out_Tot <- sum(nbetas_Out)
  
  W <- diag(n*p)
  
  # Main Loop
  IFL_t0 <- system.time({
    Ld <- L_Matrix(1, nbetas_In)
    Ld_i <- inverse_L(nbetas_In)
    
    # Update model
    H <- XX %*% W %*% Ld_i 
    
    # Ada LASSO solution with smallest BIC
    # Intercept is managed outside glmnet, therefore intercept here is FALSE
    theta_hat_Est <- Ada_LASSO_intercept(H, y,FALSE, FALSE)
    theta_hat <- theta_hat_Est$coef
    
    # VERIFICAR !!!!!
    theta_hat_Intercept <- theta_hat_Est$parameters[1]
    theta_hat_itcp <- theta_hat_Est$itcp
    # VERIFICAR !!!!!
    
    # Gamma_d
    Gamma_d_TF <- abs(theta_hat) > 0
    Gamma_d <- matrix(as.integer(Gamma_d_TF) ,
                      nrow = nbetas_In_Tot, ncol = 1, byrow = FALSE)
    Gamma_d_Acc <- apply(Gamma_d, 2, cumsum)
    
    # Checking for 1s in free betas
    aux1 <- 0  
    nbetas_In_Acc <- cumsum(nbetas_In)
    for (i in 1:p) {
      aux2 <- aux1 + min(1,nbetas_In[i]) 
      Gamma_d[(aux1+1):aux2] <- 1
      aux1 <- nbetas_In_Acc[i]
    }
    
    # Calculating qty of diff betas in the next iteration
    nbetas_Out <- c(rep(0,p))
    nbetas_Out[1] <- sum(Gamma_d[1:nbetas_In_Acc[1]])
    for (j in 2:p) {
      aux1 <- nbetas_In_Acc[j-1] + 1
      aux2 <- nbetas_In_Acc[j]
      nbetas_Out[j] = sum(Gamma_d[aux1:aux2])
    }
    nbetas_Out_Tot <- sum(nbetas_Out)
    
    # Updating Matrix M
    if(nbetas_In_Tot != nbetas_Out_Tot){  
      g_d <- cumsum(Gamma_d) 
      M <- matrix(0, nrow = nbetas_In_Tot, ncol = nbetas_Out_Tot, byrow = TRUE)
      for (i in 1:nbetas_In_Tot){
        if(Gamma_d[i] != 0){
          M[i, g_d[i]] <- 1
        } else{
          M[i, g_d[i-1]] <- 1
        }
      }
      
      # Update W
      W <- W %*% M
      nbetas_In <- nbetas_Out
      nbetas_In_Tot <- sum(nbetas_In)
      LdP <- L_Matrix(1, nbetas_In)
      LdP_i <- inverse_L(nbetas_In)
      
      # Partial solution: solve problem on new dimensions 
      HP <- XX %*% W %*% LdP_i 
      
      if (handle_outlier){
        # Handling Outliers
        HPo <- cbind(HP, diag(1, n, n))
        
        # Ada LASSO solution with smallest BIC
        theta_hatP_Est <- Ada_LASSO_intercept_outlier(HP, y, HPo, FALSE, FALSE)
        
        # Solution without outliers
        theta_hatP <- theta_hatP_Est$coef
        theta_hatp_Intercept <- theta_hatP_Est$parameters[1]
        # BIC Value
        bic_value <- theta_hatP_Est$bic[theta_hatP_Est$bick]
        
        # Solution with outliers
        theta_hatPo <- theta_hatP_Est$coefo
        theta_hatp_Intercepto <- theta_hatP_Est$itcpo
        # BIC Value
        bic_value_outliers <- theta_hatP_Est$bico[theta_hatP_Est$bicko]
        
        y_hato <- HPo %*% theta_hatPo # With Intercept
        y_hat  <- HP  %*% theta_hatP  # Without Intercept
        
        # Number of Outliers
        ind_out <- which(tail(theta_hatPo, n) > 0)
      
      } else{
        # Ada LASSO solution with smallest BIC
        theta_hatP_Est <- Ada_LASSO_intercept(HP, y, FALSE, FALSE)
        theta_hatP <- theta_hatP_Est$coef
        theta_hatp_Intercept <- theta_hatP_Est$parameters[1]
        # BIC Value
        bic_value <- theta_hatP_Est$bic[theta_hatP_Est$bick]
        bic_value_outliers <- NULL
        
        y_hato <- rep(0, n)          # With Intercept
        y_hat  <- HP  * theta_hatP  # Without Intercept
        
        #v_beta_Par <- W %*% LdP_i %*% theta_hatP
        #m_beta_Par <- matrix(v_beta_Par, nrow = n, ncol = p, byrow = FALSE)}
        
      } # Handling outliers
    } # Updating M
  }) # Time
  
  # beta estimated without outliers
  v_beta_Par <- W %*% LdP_i %*% theta_hatP # Without the intercept
  m_beta_Par <- matrix(v_beta_Par, nrow = n, ncol = p, byrow = FALSE)
  
  # beta estimated with outliers
  if (handle_outlier){
    v_beta_Paro <- W %*% LdP_i %*% theta_hatPo[1:(length(theta_hatPo)-n)]
    # Without the intercept
    m_beta_Paro <- matrix(v_beta_Paro, nrow = n, ncol = p, byrow = FALSE)
  } else{
    v_beta_Paro <- v_beta_Par
    m_beta_Paro <- m_beta_Par
  }
  
  # Estimate the intercept
  beta0_hat <- mean(y_original) - colMeans(XX) %*% v_beta_Par 
  
  # Always consider the estimated intercept. Since data was centralized, it needs to be decentralized back.
  y_original_hat <- as.numeric(beta0_hat) + y_hat 
  
  if(handle_outlier) {
    y_original_hat_woutl  <- as.numeric(beta0_hat) + y_hato
  } else{
    y_original_hat_woutl  <- y_original_hat_woutl
  }
  
  # RMSE In Sample
  RMSE_inS  <- sqrt(1/length(y)*sum((y-y_hat)^2))
  RMSE_inSo <- sqrt(1/length(y)*sum((y-y_hato)^2))
  
  # Selected variables X unselected variables, and variables with break 
  chosen_variables <- rep(0,p)
  chosen_variables_break <- rep(0,p)
  chosen_variables_outliers <- rep(0,p)
  chosen_variables_outliers_break <-  rep(0,p)
  
  for (j in 1:p) {
    if (abs(sum(m_beta_Par[,j])) > 0) {
      chosen_variables[j] <- 1
      featbk <- FALSE
      betaref <- m_beta_Par[1,j] # first value of m_beta_Par[:,j] 
      for (k in 1:n) {
        if(m_beta_Par[k,j] != betaref) { # At least one break
          featbk <- TRUE
        }
      }
      if(featbk) {
        chosen_variables_break[j] <- 1
      }
    }
    
    if (abs(sum(m_beta_Paro[,j])) > 0) {
      chosen_variables_outliers[j] <- 1
      featbk <- FALSE
      betaref <- m_beta_Paro[1,j] # first value of m_beta_Par[:,j] 
      for (k in 1:n) {
        if(m_beta_Paro[k,j] != betaref) { # At least one break
          featbk <- TRUE
        }
      }
      if(featbk) {
        chosen_variables_outliers_break[j] <- 1
      }
    }
  }
  
  list( beta_np_1 = v_beta_Par , beta_n_p = m_beta_Par, 
        betao_np_1 = v_beta_Paro , betao_n_p = m_beta_Paro,
        beta0 = beta0_hat,
        y_hat =  y_original_hat,
        y_hato = y_original_hat_woutl,
        mean_x  = mean_x , mean_y  = mean_y,
        bic_value = bic_value,
        bic_value_outliers = bic_value_outliers,
        chosen_var = chosen_variables,
        chosen_var_bk = chosen_variables_break,
        chosen_var_outliers = chosen_variables_outliers,
        chosen_var_outliers_bk = chosen_variables_outliers_break,
        RMSE_inS = RMSE_inS, RMSE_inSo = RMSE_inSo,
        ind_out = ind_out,
        IFL_t0 = IFL_t0  )
  

}
  
  

File: Julia/IFL_Algorithm.jl
#
#  IFL: Iterative Fused LASSO Algorithm
#
#  last update: AM 16/10/25
#

using LinearAlgebra
using GLMNet

# Supporting Function
include("supporting_functions.jl")

#
# IFL
#
"""
IFL(y, x; handle_outliers = true, intercept = false, CD = false, verbose = false)

Since IFL considers a dynamic model, data should be standardized OUTSIDE IFL.

CD = false # uses glmnet
CD = true  # uses Coordinate Descent
  
Returns β_np_1 = v_beta_Par , β_n_p = m_beta_Par,                 # Solutions without Outliers
        βo_np_1 = v_beta_Paro , βo_n_p = m_beta_Paro,             # Solutions with Outliers
        β0 = β0_hat, β0_hato = β0_hato,                           # Estimated Intercept
        y_hat =  y_original_hat ,                                 # Insample Estimators without Outliers
        y_hato = y_original_hat_woutl,                            # Insample Estimators with Outliers
        bic_value = bic_value,                                    # BIC value model without outliers
        bic_value_outliers = bic_value_outliers,                  # BIC value model with outliers
        chosen_var = chosen_variables,                            # List of selected columns in X
        chosen_var_bk = chosen_variables_break,                   # List of selected columns with estim breaks in X
        chosen_var_outliers = chosen_variables_outliers,          # List of selected columns in X (model w/outliers)
        chosen_var_outliers_bk = chosen_variables_outliers_break, # List of selected columns with estim breaks in X (model w/outliers)
        RMSE_inS = RMSE_inS, RMSE_inSo = RMSE_inSo,               # In Sample RMSE (models with and without outliers)
        ind_out = ind_out,                                        # list of outlier in y  
        IFL_t0 = IFL_t0                                           # time to converge

Last review: 16 Oct 25

"""
function IFL(y, x; handle_outliers = true, intercept = false, CD = false, verbose = false)
  """
  """ 
  
  # Store originals
  x_original = x
  y_original = y

  #
  mean_x = mean(x_original, dims=1)
  mean_y = mean(y_original)

  xc = x .- mean_x
  yc = y.- mean_y
    
  # Create a DLRM Intercept 
  if (intercept) 
    xc = hcat(ones(size(xc)[1]), xc)
  end
    
  ind_out=[]

  ####################################################################
  # Problem Dimensions
  n = length(y) 
  p = size(xc)[2] # When Intercept is assumed p has already turned to p-1
  ####################################################################
  
  # Setup Iteractive Fused LASSO 
  nbetas_In  = repeat([n], p)
  nbetas_In_Tot =sum(nbetas_In)
  nbetas_Out = repeat([0], p)
  nbetas_Out_Tot = sum(nbetas_Out)

  global XX = Matrix{Float64}(undef, 0, 0)  # Alterei de Int para Float64!!! 27/05/25
  for i in 1:p
      if (size(XX, 1) == 0)
          XX = Diagonal(xc[:, i])  # Initialize with the first diagonal matrix
      else
          XX = hcat(XX, Diagonal(xc[:, i]))  # Horizontally concatenate diagonal matrices
      end
  end

  # Main loop
  IFL_t0 = @elapsed begin 
    
    time_H = @elapsed begin
      Hs = build_H(xc)
    end
    if(verbose)
      println("Step 1, Build H: ", time_H)
    end 
    
    H = Matrix(Hs)         

    # Ada LASSO solution with smallest BIC
    # Intercept is being managed outside glmnet. 
    # Since its a Dynamic Linear Regression, the intercept is time variant.
    # GLMNet consider the intercept only for obs #1, therefore intercept here is always FALSE.
    time_AdaS1 = @elapsed begin
      theta_hat_Est = Ada_LASSO_intercept(H, yc, standardize = false, intercept = false, CD = CD) 
    end 
    if(verbose)
      println("Step 1, Ada Step1: ", time_AdaS1)
    end 
    theta_hat = theta_hat_Est.coef
    
    # Gamma_d
    Gamma_d = abs.(theta_hat) .> 0
    #Gamma_d_Acc = cumsum(Gamma_d) 
      
    # Checking for 1s in free betas
    aux1 = 0
    nbetas_In_Acc = cumsum(nbetas_In)
    for i in 1:p
      aux2 = aux1 + min(1,nbetas_In[i]) 
      Gamma_d[(aux1+1):aux2] .= 1
      aux1 = nbetas_In_Acc[i]
    end
      
    # Calculating qty of diff betas in the next iteration
    nbetas_Out = zeros(p)
    nbetas_Out[1] = sum(Gamma_d[1:nbetas_In_Acc[1]])
    for j in 2:p
      aux1 = nbetas_In_Acc[j-1] + 1
      aux2 = nbetas_In_Acc[j]
      nbetas_Out[j] = sum(Gamma_d[aux1:aux2])
    end
    nbetas_Out_Tot = sum(nbetas_Out)
      
    # Updating Matrix M
    if(nbetas_In_Tot != nbetas_Out_Tot)  
      g_d = cumsum(Gamma_d) 
      M = zeros(Int(nbetas_In_Tot), Int(nbetas_Out_Tot))

      time_M = @elapsed begin   
        for i in 1:nbetas_In_Tot
          if(Gamma_d[i] != 0)
            M[i, g_d[i]] = 1
          else
            M[i, g_d[i-1]] = 1
          end
        end
      end
      if(verbose)
        println("Step 2, matrix M: ", time_M)
      end

      # Update W = M
      nbetas_In = nbetas_Out
      nbetas_In_Tot = sum(nbetas_In)
      
      global LdP_i = Inverse_L(Int.(nbetas_In))
      # Partial solution: solve problem on new dimensions 
      global HP = XX * M * LdP_i 
        
      if (handle_outliers) 
        # Handlig Outliers
        global HPo = hcat(HP, Matrix{Float64}(I, Int(n), Int(n)))
            
        # Ada LASSO solution with smallest BIC
        # Intercept is being managed outside glmnet, therefore intercept here is always FALSE
        time_AdaS3wO = @elapsed begin   
          theta_hatP_Est = Ada_LASSO_intercept_outliers(HP, yc, HPo, standardize = false, intercept = false, CD = CD) 
        end
        if(verbose)
          println("Step 3, AdaLASSO with Outliers: ", time_AdaS3wO)
        end
        
        # Solution without outliers
        global theta_hatP = theta_hatP_Est.coef
        global theta_hatP_itcp = theta_hatP_Est.itcp
        # BIC value
        global bic_value = theta_hatP_Est.bic[theta_hatP_Est.bick] 

        # Solution with Outliers
        global theta_hatPo = theta_hatP_Est.coefo
        #global theta_hatPo_itcpo = theta_hatP_Est.itcpo
        # BIC value
        global bic_value_outliers = theta_hatP_Est.bico[theta_hatP_Est.bicko] 
          
        global y_hato = HPo * theta_hatPo           # Without Intercept 
        global y_hat  = HP * theta_hatP             # Without Intercept
          
        # No. of Outliers
        ind_out = findall(pts -> pts != 0, theta_hatPo[(end-n+1):end])  
                  
      else
        time_AdaS3woO = @elapsed begin
          theta_hatP_Est = Ada_LASSO_intercept(HP, yc, standardize = false, intercept = false, CD = CD) 
        end
        if(verbose)
          println("Step 3, AdaLASSO without Outliers: ", time_AdaS3woO)
        end
        
        global theta_hatP = theta_hatP_Est.coef
        #global theta_hatP_itcp = theta_hatP_Est.itcp
        # BIC value
        global bic_value = theta_hatP_Est.bic[theta_hatP_Est.bick] 
        global bic_value_outliers = nothing 
        
        global y_hat = HP * theta_hatP # Without Intercept
        global y_hato = zeros(n) 
          
      end # Handle Outliers
    end # Updating M
  end  # Time

  # β estimated without outliers
  v_beta_Par = M * LdP_i * theta_hatP     
  m_beta_Par = reshape(v_beta_Par, n, p) 

  # β estimated with outliers
  if(handle_outliers)                 
    v_beta_Paro = M * LdP_i * theta_hatPo[1:end-n] # dim(theta_hatPo) is (np + n) times 1
    # Without Intercept
    m_beta_Paro = reshape(v_beta_Paro, n, p) 
  else # without outliers => always returns a value for v_beta and m_beta 
    v_beta_Paro = v_beta_Par
    m_beta_Paro = m_beta_Par 
  end
    
  # Estimate Intercept
  #if (intercept)
  #  mean_x = hcat(1, mean_x) 
  #end
  #β0_hat  = mean_y .- mean_x  * m_beta_Par[end,:]  # valid only for the last intercept.
  #β0_hato = mean_y .- mean_x  * m_beta_Paro[end,:] # valid only for the last intercept. 
  if(intercept) # Estimated Intercept is the first colum of B = m_beta_Par
    β0_hat = m_beta_Par[:,1]  
    β0_hato = m_beta_Paro[:,1]  
    m_beta_Par = m_beta_Par[:,2:end]
    m_beta_Paro = m_beta_Paro[:,2:end]
    p = p-1
    v_beta_Par = reshape(m_beta_Par, n*p, 1)
    v_beta_Paro = reshape(m_beta_Paro, n*p, 1)
  else
    β0_hat  = mean_y .- mean_x  * m_beta_Par[end,:]  # valid only for the last intercept.
    β0_hato = mean_y .- mean_x  * m_beta_Paro[end,:] # valid only for the last intercept. 
  end
      
  # Since data was centralized, it needs to be decentralized back.
  ydc_hat = y_hat .+ mean_y
  if(handle_outliers)
    ydc_hat_wOutl  = y_hato .+ mean_y
  end
    
  # RMSE In Sample
  RMSE_inS  = RMSE(y,ydc_hat) 
  RMSE_inSo = RMSE(y, ydc_hat_wOutl) 
  
  # Selected variables X unselected variables, and variables with break 
  chosen_variables = zeros(p)
  chosen_variables_break =  zeros(p)
  chosen_variables_outliers = zeros(p)
  chosen_variables_outliers_break =  zeros(p)
  
  # p = number of original variables (does not consider intercept) 
  for j in 1:p
    if (abs.(sum(m_beta_Par[:,j])) > 0)
      chosen_variables[j] = 1
      featbk = false
      βref = m_beta_Par[1,j] # first value of m_beta_Par[:,j] 
      for k in 1:n
        if(m_beta_Par[k,j] != βref) # At least one break
            featbk = true
        end
      end
      if(featbk)
        chosen_variables_break[j] = 1
      end
    end
    if (abs.(sum(m_beta_Paro[:,j])) > 0)
      chosen_variables_outliers[j] = 1
      featbk = false
      βref = m_beta_Paro[1,j] # first value of m_beta_Par[:,j] 
      for k in 1:n
        if(m_beta_Paro[k,j] != βref) # At least one break
            featbk = true
        end
      end
      if(featbk)
        chosen_variables_outliers_break[j] = 1
      end
    end
  end
  
  return (β_np_1 = v_beta_Par , β_n_p = m_beta_Par,
          βo_np_1 = v_beta_Paro , βo_n_p = m_beta_Paro,
          β0_hat = β0_hat, β0_hato = β0_hato,
          y_hat =  ydc_hat ,  y_hato = ydc_hat_wOutl,
          bic_value = bic_value, bic_value_outliers = bic_value_outliers,            
          chosen_var = chosen_variables,
          chosen_var_bk = chosen_variables_break,
          chosen_var_outliers = chosen_variables_outliers,
          chosen_var_outliers_bk = chosen_variables_outliers_break,
          RMSE_inS = RMSE_inS, RMSE_inSo = RMSE_inSo,
          ind_out = ind_out, IFL_t0 = IFL_t0)

end
